syntax = "proto3";
package venom;

message Packet {
    enum Type {
        UNKNOWN = 0;
        REGISTER = 1;
        LOOKUP_USER = 2;       // By Username (@cv2 -> Key)
        LOOKUP_USER_BY_ID = 3; // By Key (Key -> @cv2) <--- NEW
        SEND = 4;
        FETCH = 5;
    }

    Type type = 1;

    // Auth
    bytes sender_id_pubkey = 2;
    bytes signature = 3;
    uint64 timestamp = 4;

    RegisterPayload register = 5;
    string lookup_username = 6;
    bytes lookup_user_id = 7;     // <--- NEW: For LOOKUP_USER_BY_ID

    // Routing
    bytes target_id_pubkey = 8;
    Envelope envelope = 9;
}

message RegisterPayload {
    string username = 1;        // "cv2"
    bytes enc_pubkey = 2;       // X25519 Key (Public Directory needs this)
}

message Response {
    uint32 status = 1;       // 200 OK, 409 Conflict (User exists), 404 Not Found
    string error_msg = 2;

    // Return Data
    UserInfo user_info = 3;             // Result of LOOKUP
    repeated Envelope pending_messages = 4; // Result of FETCH
}

message UserInfo {
    string username = 1;
    bytes id_pubkey = 2;   // Ed25519
    bytes enc_pubkey = 3;  // X25519
}

// ... Envelope, Payload remain same ...
message Envelope {
    bytes sender_identity_key = 1;
    bytes ephemeral_pubkey = 2;
    bytes nonce = 3;
    bytes ciphertext = 4;
    bytes signature = 5;
}

message Payload {
    enum Type { UNKNOWN=0; TEXT=1; }
    Type type = 1;
    uint64 timestamp = 2;
    string body = 3;
}